### PR TITLE
rework on the trailing whitespace script

### PR DESCRIPTION
**Pull request by srajangarg**
Have made it a bit cleaner. Also made it more general, if more checks need to be added later.
@Sumith1896 Maybe rename the file to `check_code_quality`?

- [x] clean up file, and make more OO
- [ ] add new "spaces after arguments" check
- [ ] add "spaces before and after '=' or '=='" check

### CODE REVIEW

**Comment by Sumith1896** [comment]
Thanks! Having a look at it.

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 1]
```diff
@@ -1,98 +1,72 @@
...
-    """
-    Checks all files in the directory tree (with base_path as starting point)
-    with the file_check function provided, skipping files that contain
-    any of the strings in the set provided by exclusions.
-    """
-    if not base_path:
-        return
-    for root, dirs, files in walk(base_path):
-        check_files(glob(join(root, pattern)), file_check, exclusions)
+class Check:
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: delete

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 2]
```diff
@@ -1,98 +1,72 @@
...
+        self.excluded_names = excluded_names
+
+
+
+######                 CONSTRUCT YOUR CHECK CLASSES HERE            ######
+all_checks = []
+
+# trailing whitespace check
+def trailing_whitespace_function(file_path):
+
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: delete extra line

**Comment by Sumith1896** [comment]
Also @srajangarg I think this script was not running on Travis after it was moved. Hence, some trailing whitespaces have crept after that. Could you remove them along with this PR?

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 3]
```diff
@@ -1,98 +1,72 @@
...
-        for i,failure in enumerate(_failed_expectations, start=1):
-            report.append('%d: %s' % (i, failure))
-        _failed_expectations = []
-    if len(report) != 0:
-        return ('\n'.join(report))
-
-SYMENGINE_PATH = abspath(join(split(__file__)[0], pardir, pardir))  # go to symengine/
-assert exists(SYMENGINE_PATH)
+from os import walk, pardir
+from os.path import split, join, abspath, exists
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: delete this line

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 4]
```diff
@@ -1,98 +1,72 @@
...
+
+    Arguments
+    =========
+    checker_fx          : [function] takes in absolute filepath as argument and does the checking
+    root_dirs           : [list] directories on which the check will run recursively
+    included_filetypes  : [optional] [list] patterns, the file will be matched against
+    excluded_names      : [optional] [list] file won't be checked, if it's abs path is `in` excluded_names
+
     """
+    def __init__(self, checker_fx, root_dirs, included_filetypes=["."], excluded_names=[]):
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: new line before this

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 5]
```diff
@@ -1,98 +1,72 @@
...
+
+        self.fx = checker_fx
+        self.root_dirs = root_dirs
+        self.included_filetypes = included_filetypes
+        self.excluded_names = excluded_names
+
+
+
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: delete one line of the three.

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 6]
```diff
@@ -1,98 +1,72 @@
...
+######                 CONSTRUCT YOUR CHECK CLASSES HERE            ######
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: `# Construct your check classes here`

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 7]
```diff
@@ -1,98 +1,72 @@
...
+all_checks = []
+
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: define the bool `errors` here and then use it later

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 8]
```diff
@@ -1,98 +1,72 @@
...
+# trailing whitespace check
+def trailing_whitespace_function(file_path):
+
+    global Errors
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: convert to `errors`

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 9]
```diff
@@ -1,98 +1,72 @@
...
 
-def test(fname):
+trailing_whitespace_check = Check(trailing_whitespace_function, [SYMENGINE_PATH])
+trailing_whitespace_check.excluded_names = ["/build/", "/doc/", "/cmake/", "/utilities"]
+trailing_whitespace_check.included_filetypes = [".cpp", ".h", ".py", ".sh"]
+all_checks.append(trailing_whitespace_check)
 
-    with open(fname, "rt") as test_file:
-        test_this_file(fname, test_file)
+# another check
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: `# Add another check below`

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 10]
```diff
@@ -1,98 +1,72 @@
...
-def test_this_file(fname, test_file):
-    line = None  # to flag the case where there were no lines in file
-    tests = 0
-    test_set = set()
-    for idx, line in enumerate(test_file):
-        if line.endswith(" \n") or line.endswith("\t\n"):
-            _log_failure( message_space % (fname, idx + 1))
+########                         MAIN                              #######
+Errors = False
+for check in all_checks:
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: remove

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 11]
```diff
@@ -1,98 +1,72 @@
...
 
-exclude = set()
+    for base_dir in check.root_dirs:
+        for root, dirs, file_names in walk(base_dir):
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: remove

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 12]
```diff
@@ -1,98 +1,72 @@
...
 
-check_directory_tree(BIN_PATH, test, set(["~",".sh"]), "*")
-check_directory_tree(SYMENGINE_PATH, test, set(["/build/","/doc/","/cmake/","/utilities"]))
-check_directory_tree(SYMENGINE_PATH, test, set(["/build/","/doc/","/cmake/","/utilities"]), "*.h")
+            for file_name in file_names:
+
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: remove

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 13]
```diff
@@ -1,98 +1,72 @@
...
-exclude = set()
+    for base_dir in check.root_dirs:
+        for root, dirs, file_names in walk(base_dir):
 
-check_directory_tree(BIN_PATH, test, set(["~",".sh"]), "*")
-check_directory_tree(SYMENGINE_PATH, test, set(["/build/","/doc/","/cmake/","/utilities"]))
-check_directory_tree(SYMENGINE_PATH, test, set(["/build/","/doc/","/cmake/","/utilities"]), "*.h")
+            for file_name in file_names:
+                
+                file_path = join(root, file_name)
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: remove

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 14]
```diff
@@ -1,98 +1,72 @@
...
 
-print _report_failures()
+                if any(excluded_name in file_path for excluded_name in check.excluded_names):
+                    continue
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: remove

**Comment by Sumith1896** [comment]
Minor nits here and there, apart from that, I like this a lot. Much cleaner. Thanks @srajangarg

**Comment by Sumith1896** [comment]
Yes you can rename the file since you have a general construct. Maybe `test_code_quality.py` like in SymPy?

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 15]
```diff
@@ -1,98 +1,72 @@
...
-        source = test_file.read()
-	
-def test_this_file(fname, test_file):
-    line = None  # to flag the case where there were no lines in file
-    tests = 0
-    test_set = set()
-    for idx, line in enumerate(test_file):
-        if line.endswith(" \n") or line.endswith("\t\n"):
-            _log_failure( message_space % (fname, idx + 1))
+########                         MAIN                              #######
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: `# Main`

File: symengine/utilities/tests/test_trailing_whitespace.py [hunk 16]
```diff
@@ -1,98 +1,64 @@
...
 
-print _report_failures()
+                if any(excluded_name in file_path for excluded_name in check.excluded_names):
+                    continue
 
-if len(report) != 0:
-    raise Exception('Trailing whitespaces present')
+                if any(file_path.endswith(file_type) for file_type in check.included_filetypes):
+                    check.fx(file_path)
+if errors:
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_trailing_whitespace.py]: new line before this

**Comment by Sumith1896** [comment]
Thanks @srajangarg. Looks good to me.
Can merge once Travis passes.

**Comment by CodeMaxx** [comment]
@srajangarg  variable BIN_PATH is not used in the script. So [this](https://github.com/symengine/symengine/pull/733/files#diff-fa51d6e51ab12dc352fed6e80b42112eR7) can be removed.

File: symengine/utilities/tests/test_code_quality.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
+from os import walk, pardir
+from os.path import split, join, abspath, exists
+
+### CHANGE THIS IF THIS FILE IS EVER MOVED 
+### RIGHT NOW IT'S IN symengine/symengine/utilities/tests/
+SYMENGINE_PATH = abspath(join(split(__file__)[0], pardir, pardir))
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_code_quality.py]: Need to go one directory back to include `bin` and `benchmarks`

File: symengine/utilities/tests/test_code_quality.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+def trailing_whitespace_function(file_path):
+    global errors
+    with open(file_path) as file_ptr:
+        for line_num, line in enumerate(file_ptr):
+            if line.endswith(" \n") or line.endswith("\t\n"):
+                print "Trailing whitespace :", file_path, str(line_num + 1)
+                errors = True
+
+trailing_whitespace_check = Check(trailing_whitespace_function, [SYMENGINE_PATH])
+trailing_whitespace_check.excluded_names = ["/build/", "/doc/", "/cmake/", "/utilities"]
```
Comment (Sumith1896) [inline symengine/utilities/tests/test_code_quality.py]: can remove `/utilities` and `/cmake`

File: symengine/utilities/tests/test_code_quality.py (hunk 17 above)
Comment (CodeMaxx) [inline symengine/utilities/tests/test_code_quality.py]: Correct . So we need another variable benchmark_path

Comment (Sumith1896) [inline symengine/utilities/tests/test_code_quality.py]: else just append a `pardir`

File: symengine/utilities/tests/test_code_quality.py (hunk 18 above)
Comment (CodeMaxx) [inline symengine/utilities/tests/test_code_quality.py]: We have `conf.py` in `/doc/` . Add a test for that too , incase that file is changed in the future.

Comment (Sumith1896) [inline symengine/utilities/tests/test_code_quality.py]: Let's skip doc for now.

**Comment by Sumith1896** [comment]
@srajangarg The new TODOs you have put is interesting but I think it can go in a different PR. Let's this get merged first. 

@isuruf Could you have a look?

**Comment by CodeMaxx** [comment]
@srajangarg  is the skipping of `benchmarks` directory intentional?

**Comment by Sumith1896** [comment]
Sorry @srajangarg for the multiple merge conflicts you had to resolve. 
I think this can be merged.
@certik or @isuruf will do so when they see this.

**Comment by certik** [comment]
That looks good to me. Thanks @srajangarg !

**Comment by certik** [comment]
Thanks @Sumith1896 and @CodeMaxx for reviewing this.

**Comment by Sumith1896** [comment]
Thanks @srajangarg . If you work on the others, ping me on that PR.

**Pull request merged**